platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemReport.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.problems.internal;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.problems.Severity;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@NonNullApi
+public class DefaultProblemReport implements Serializable, ProblemReport {
+    private final ProblemDefinition problemDefinition;
+    private final ProblemContext problemContext;
+
+
+    protected DefaultProblemReport(
+        String label,
+        Severity severity,
+        List<ProblemLocation> locations,
+        @Nullable DocLink documentationUrl,
+        @Nullable String description,
+        @Nullable List<String> solutions,
+        @Nullable RuntimeException cause,
+        ProblemCategory problemCategory,
+        Map<String, Object> additionalData
+    ) {
+        this.problemDefinition = new DefaultProblemDefinition(label, severity, documentationUrl, solutions, problemCategory);
+        this.problemContext = new DefaultProblemContext(locations, description, cause, additionalData);
+    }
+
+    @Override
+    public ProblemDefinition getProblemDefinition() {
+        return problemDefinition;
+    }
+
+    @Override
+    public ProblemContext getProblemContext() {
+        return problemContext;
+    }
+
+    @Override
+    public InternalProblemBuilder toBuilder() {
+        return new DefaultProblemBuilder(this);
+    }
+
+    private static boolean equals(@Nullable Object a, @Nullable Object b) {
+        return (a == b) || (a != null && a.equals(b));
+    }
```

Comment:
🤔 Just my guess, but I guess this was made, as `Objects.equals(...)` is only available 1.7+?

Reply:
Yes indeed :)